app/frontend/wallet/helpers/CachedDeriveXpubFactory.ts

```diff
@@ -28,14 +37,55 @@ function CachedDeriveXpubFactory(derivationScheme, deriveXpubHardenedFn) {
     return await derivedXpubs[memoKey]
   }
 
-  async function deriveXpubNonhardenedFn(derivationPath) {
+  async function deriveXpubNonhardenedFn(derivationPath: BIP32Path) {
     const lastIndex = derivationPath.slice(-1)[0]
     const parentXpub = await deriveXpub(derivationPath.slice(0, -1))
-
     return deriveChildXpub(parentXpub, lastIndex, derivationScheme.ed25519Mode)
   }
 
-  return deriveXpub
+  function createPathBulk(derivationPath: BIP32Path): BIP32Path[] {
+    const paths: BIP32Path[] = []
+    const accountIndex = derivationPath[2] - HARDENED_THRESHOLD
+    const currentAccountPage = Math.floor(accountIndex / MAX_BULK_EXPORT_AMOUNT)
+
+    /*
+    * in case of the account 0 we append also the byron path
+    * since during byron era only the first account was used
+    */
+    if (accountIndex === 0) paths.push(BYRON_V2_PATH)
+
+    for (let i = 0; i < MAX_BULK_EXPORT_AMOUNT; i += 1) {
+      const nextAccountIndex = currentAccountPage * MAX_BULK_EXPORT_AMOUNT + i + HARDENED_THRESHOLD
+      const nextAccountPath = [...derivationPath.slice(0, -1), nextAccountIndex]
+      paths.push(nextAccountPath)
+    }
+    return paths
+  }
+
+  async function deriveXpubHardenedFn(derivationPath: BIP32Path): Promise<any> {
+    const paths = shouldExportPubKeyBulk ? createPathBulk(derivationPath) : [derivationPath]
+    const xPubBulk = await deriveXpubFn(paths)
```

Comment:
do we have to await here? Can't we just take the xPubBulk promise and do below:
```
      _derivedXpubs[memoKey] = xPubBulk.then((keys) => return keys[i]).reject(() => return null)
```

would we need the cleanXpubCache function then?